### ISSUE TITLE
Memory leak running on lightweight model

### ISSUE THREAD

**Issue by innocentius** [unitaryai/detoxify#7]
First of all, this repository has been a great help for my research effort, and I really appreciate you sharing this to the public.

My issue is that the program would not free up its memory after done with the calculation. Here is the code to replicate the issue:

`from detoxify import Detoxify`
`import torch`
`Detoxify('original-small').predict("Beep, beep, I'm a sheep.")`
`torch.cuda.empty_cache()`

When monitoring GPU memory use, we find that the memory is not freed (about 600M more than before). This could be a problem when we run the model with large amount of data or multithreading. I have been trying to reasonably multithreading the model, only to find it always run out of memory when running larger datasets (each thread runs the model only one time then joins, 5 threads are used but after running 500 times it exceed 8G of GPU memory). Only after the kernel is completely destroyed the memory would be freed. I'm using Windows 10 Anaconda 1.10, Python 3.8.3 , PyTorch 1.7 w/ CUDA 11.0.

I'm currently looking into ways to solve this, would you be able to help?

PS: The same issue could present in original model, maybe GPU and CPU computing uses different garbage collection methods.

**Comment by jamt9000** [comment]
Thanks for the report.

Can you check if the cuda memory is freed when the Detoxify object is explicitly deleted?

```.py
from detoxify import Detoxify
model = Detoxify('original-small')
model.predict("Beep, beep, I'm a sheep.")
del model
```

**Comment by innocentius** [comment]
```python
import torch
from detoxify import Detoxify
model = Detoxify('original-small')  #About 500 MB increase in memory use
model.predict("Beep, beep, I'm a sheep.") #no increase seen
del model #no memory decrease seen
torch.cuda.empty_cache() #-100mb in memory use
 ```

Here is what I see with a single case. However, It seems that in my code I always create a new detoxify('original-small') model every time I try to analyze something, like this:

```python

def Detoxifier(strin):
    return detox_model.predict(str(strin))

``` 

So it seems that guiding people to create a global varilable for the model might be a good idea.

**Comment by jamt9000** [comment]
I see, yes you shouldn't instantiate the model from scratch for each prediction because it will load the weights from disk each time making things slow, we'll make that clearer. It doesn't need to be a global variable, you can just instantiate it once at the top level scope.

That said, I'd still expect the memory to be reclaimed at some point once the Detoxify instance is deleted.

Does the memory get freed if you do `import gc; gc.collect()` to manually run the python garbage collector?

If you're using Jupyter notebooks another issue can be that cell outputs are kept around indefinitely in the `Out[n]` list so references to large objects can be subtly kept around.

**Comment by innocentius** [comment]
Running gc doesn't help with the GPU memory situation, at least in this case. I would consider the possibility that some of the variables in the model is declared global and couldn't be reclaimed even if the detoxify instance is erased. 
I'm using Spyder, and I'm not sure if the kernel outputs are treated differently here.

**Comment by jamt9000** [comment]
Ok, I'll close this issue since it seems to be solved by only initialising the model once. Thanks for your help debugging! @laurahanu can you update the docs to make it clearer that `Detoxify('original-small').predict('asdf')` initialises the model each time?
